README.md
# FVTT Token Vision Tweaks
Module for Foundry VTT that provides some tweaks for Token Vision, to allow trading quality for performance, or vice-versa.

⚠ This module is not needed if you're running Foundry 0.8 or newer.

[![License](https://img.shields.io/github/license/ruipin/fvtt-token-vision-tweaks)](LICENSE)
[![Build Release](https://github.com/ruipin/fvtt-token-vision-tweaks/workflows/Build%20Release/badge.svg)](https://github.com/ruipin/fvtt-token-vision-tweaks/releases/latest)
[![Version (latest)](https://img.shields.io/github/v/release/ruipin/fvtt-token-vision-tweaks)](https://github.com/ruipin/fvtt-token-vision-tweaks/releases/latest)
[![Foundry Version](https://img.shields.io/badge/dynamic/json.svg?url=https://github.com/ruipin/fvtt-token-vision-tweaks/releases/latest/download/module.json&label=Foundry%20Version&query=$.compatibleCoreVersion&colorB=blueviolet)](https://github.com/ruipin/fvtt-token-vision-tweaks/releases/latest)
[![GitHub downloads (latest)](https://img.shields.io/badge/dynamic/json?label=Downloads@latest&query=assets[?(@.name.includes('zip'))].download_count&url=https://api.github.com/repos/ruipin/fvtt-token-vision-tweaks/releases/latest&color=green)](https://github.com/ruipin/fvtt-token-vision-tweaks/releases/latest)
[![Forge Install Base](https://img.shields.io/badge/dynamic/json?label=Forge%20Install%20Base&query=package.installs&suffix=%&url=https://forge-vtt.com/api/bazaar/package/token-vision-tweaks&colorB=brightgreen)](https://forge-vtt.com/)
[![GitHub issues](https://img.shields.io/github/issues-raw/ruipin/fvtt-token-vision-tweaks)](https://github.com/ruipin/fvtt-token-vision-tweaks/issues)
[![Ko-fi](https://img.shields.io/badge/-buy%20me%20a%20coffee-%23FF5E5B?logo=Ko-fi&logoColor=white)](https://ko-fi.com/ruipin)


## Features

* Allows configuration of the ray deduplication tolerance, both globally or per-scene. This option can be used to massively improve vision on large open maps at a cost of performance.

* Allows configuration of the ray cast density for Token sight, both globally or per-scene. This option can be used to slightly improve vision on large open maps at a cost of performance, although in most cases increasing the 'Ray De-duplication Tolerance' is much more effective.

* Allows setting the maximum token vision distance, regardless of illumination settings, both globally or per-scene. This can be used to e.g. limit view distance when Global Illumination is turned on, or improve performance on very large open maps by limiting how far away tokens need to render sight.

* Allows configuration of Foundry's 'Exact Vision Threshold', both globally or per-scene. This is the number of walls in the map at which vision becomes approximate.


## Installation
1. Copy this link and use it in Foundry's Module Manager to install the Module

    > https://github.com/ruipin/fvtt-token-vision-tweaks/releases/latest/download/module.json

2. Enable the Module in your World's Module Settings

### libWrapper

This module uses the [libWrapper](https://github.com/ruipin/fvtt-lib-wrapper) library for wrapping core methods. While this is not a hard dependency, it is recommended to install it for the best experience and compatibility with other modules.

CHANGELOG.md
# 1.3.2.0 (2021-05-24)

* Update libWrapper shim to latest version.
* Do nothing and warn user if running under Foundry 0.8 or newer.

# 1.3.1.4 (2020-12-30)

* No code changes.
* Explicitly announce compatibility with Foundry 0.7.9.

# 1.3.1.3 (2020-12-08)

* No code changes.
* Explicitly announce compatibility with Foundry 0.7.8.

# 1.3.1.2 (2020-11-15)

* No code changes.
* Explicitly announce compatibility with Foundry 0.7.7.

# 1.3.1 (2020-10-22)

* Allow changing Foundry's 'Exact Vision Threshold'. Higher values give more accurate vision, at a cost of performance.

# 1.3.0 (2020-10-22)

* Add support for Foundry 0.7.5.
* Remove support for Foundry 0.6.x.
* Remove options that applied only to 0.6.x and are no longer relevant for 0.7.5.
* Add an option to change the Ray De-duplication Tolerance. This essentially replaces the "Aggressive Culling Fix" in the previous version of the module, and can be used to massively improve vision accuracy in large open maps.
* Remove the option for GM Fog-of-War to only update when player-controlled tokens are moved. This option was useful, but would take a lot of effort to update for 0.7.5.

# 1.2.2 (2020-08-05)

* Use libWrapper instead of a custom wrapper library.

# 1.2.1 (2020-07-29)

* Updated wrapper library, in order to improve compatibility with more modules.

# 1.2.0 (2020-07-29)

* Code cleanup.
* Massively increased compatibility with other modules.
* Added Fog-of-War memory leak hotfix.
*Fixed bug where 'Only Player-Visible Tokens Update GM Fog of War' wouldn't turn on in some cases.

# 1.1.0 (2020-07-28)

* Added 'Only Player-Visible Tokens Update GM Fog of War' setting

# 1.0.0 (2020-07-27)

* First release.

main.js
// SPDX-License-Identifier: GPLv3-or-later
// Copyright © 2020 fvtt-token-vision-tweaks Rui Pinheiro

'use strict';

Hooks.once('ready', () => {
	const MODULE_NAME = "Token Vision Tweaks";
	const MODULE_ID = "token-vision-tweaks";

	const ORIGINAL_RAY_DENSITY = 6; // from SightLayer.computeSight


	console.log(`Loading ${MODULE_NAME} module...`);


	//---------------------------
	// Settings
	const resetSight = function() {
		// Re-initialize all tokens in the scene
		// TODO: Only re-initialize the controlled tokens?
		canvas.sight.initializeTokens();
	};

	game.settings.register(MODULE_ID, 'fix-aggressive-wall-culling', {
		name: 'Fix Aggressive Wall Culling',
		default: false,
		type: Boolean,
		scope: 'world',
		scene: false,
		config: true,
		hint: "Foundry VTT 0.6.5 and 0.7.0 start culling walls at 50 grid spaces distance from a token, no matter how far it can see. This seems to be a bug, and can cause vision to phase through walls on large open maps. Checking this option will disable this aggressive culling behavior, massively improving vision on such maps, though possibly at cost of performance.",
		onChange: value => resetSight()
	});

	game.settings.register(MODULE_ID, 'ray-density', {
		name: 'Ray Density',
		default: 0,
		type: Number,
		scope: 'world',
		scene: true,
		config: true,
		hint: `The density in degrees to which to guarantee that rays are broadcast. This can be used to trade performance for quality, or vice-versa. For example, wide open maps with few walls might look better with a higher ray density, while small maps with lots of walls might perform better with a lower ray density without a noticeable quality change. Lower is better, but also slower. The Foundry default is '${ORIGINAL_RAY_DENSITY}'. Disable this override with '0'.`,
		onChange: value => resetSight()
	});

	game.settings.register(MODULE_ID, 'max-radius', {
		name: 'Maximum Vision Distance (Grid Units)',
		default: 0,
		type: Number,
		scope: 'world',
		scene: true,
		config: true,
		hint: "Maximum token vision distance in grid units, regardless of illumination settings. This can be used to e.g. limit view distance when Global Illumination is turned on, or improve performance on very large open maps by limiting how far away tokens need to render sight. Disable this override with '0'.",
		onChange: value => resetSight()
	});

	game.settings.register(MODULE_ID, 'fow-only-player-visible', {
		name: 'Only Player Tokens Update GM Fog of War',
		default: false,
		type: Boolean,
		scope: 'world',
		config: true,
		hint: "If set, the GM Fog of War will only be updated by tokens where there is at least one non-GM player with 'Owner' or 'Observer' permissions. Default Foundry behaviour is that any token with vision updates the GM Fog of War, even NPCs."
	});

	game.settings.register(MODULE_ID, 'fix-fow-memory-leak', {
		name: 'Fix Fog-of-War Memory Leak',
		default: false,
		type: Boolean,
		scope: 'world',
		config: true,
		hint: "Foundry VTT 0.6.5 and 0.7.0 have a memory leak bug involving fog of war, when 'updateToken' events are sent repeatedly without player's tokens exploring new regions of the map. In very particular cases, this can result in an Out-of-Memory crash. Checking this option applies a hotfix that prevents this issue from occurring."
	});

	const getSetting = function(key) {
		let result = game.scenes.active.getFlag(MODULE_ID, key);

		if(result === null || result === undefined)
			result = game.settings.get(MODULE_ID, key);

		return result;
	};


	//---------------------------
	// Add settings to the SceneConfig dialog
	Hooks.on('renderSceneConfig', (app, html, data) => {
		const selector = html.find('h3:contains(Vision)').nextAll('hr')[0];

		game.settings.settings.forEach(setting => {
			if(setting.module != MODULE_ID)
				return;

			if(!setting.scene)
				return;

			let value = app.object.getFlag(MODULE_ID, setting.key);
			if(value === undefined || value === null)
				value = "";

			$(`
				<div class="form-group">
					<label>${setting.name}</label>
					<input type="text" name="${setting.key}" placeholder="${setting.type.name}" value="${value}" data-dtype="${setting.type.name}"/>
					<p class="notes">${setting.hint} Leave empty to use value set in the ${MODULE_NAME} module settings.</p>
				</div>
			`).insertBefore(selector);
		});
	});

	Hooks.on('preUpdateScene', (entity, diff, options, user_id) => {
		game.settings.settings.forEach(setting => {
			if(setting.module != MODULE_ID)
				return;

			if(!setting.scene)
				return;

			if(!(setting.key in diff))
				return;

			let value = diff[setting.key];
			if(value == "null")
				value = null;

			entity.setFlag(MODULE_ID, setting.key, value);
		});
	});


	Hooks.on('updateScene', (entity, diff, options, user_id) => {
		if(!diff.flags)
			return;

		if(!diff.flags[MODULE_ID])
			return;

		resetSight();
	});


	//---------------------------
	// Hook the SightLayer computeSight method and implement the main module functionality
	new ResilientWrapper(SightLayer, 'computeSight', (function () {
		return function(computeSight, ...args) {
			let options = args[2];

			// Ray Density
			let rayDensity = getSetting('ray-density');

			if(rayDensity && rayDensity > 0 && rayDensity != ORIGINAL_RAY_DENSITY)
				options.density = rayDensity;


			// Sight Radius
			let maxRadius = (getSetting('max-radius') * canvas.dimensions.size) / canvas.dimensions.distance;

			if(maxRadius && maxRadius > 0)
				args[1] = Math.min(args[1], maxRadius);


			// Disable aggressive wall culling
			if(getSetting('fix-aggressive-wall-culling')) {
				if('cullMin' in options)
					options.cullMax = Math.max(options.cullMax, options.cullMin);
			}

			// Call wrapped method
			return computeSight.apply(this, args);
		};
	})());



	//---------------------------
	// Do not update GM FoW for tokens not owned/observed by any player
	Token.prototype.playerObservable = function() {
		if(this.actor != null) {
			if(this.actor.isPC)
				return true;

			let permissions = this.actor.data.permission;

			for( let user_id in permissions ) {
				if(user_id == "default" || user_id == game.user._id)
					continue;

				let user = game.users.find(u => {return u._id == user_id});

				if(!user || user.isGM)
					continue;

				if(permissions[user_id] >= 3) {
					return true;
				}
			}

			return false;
		}
	};


	(function() {
		let token = null;

		new ResilientWrapper(SightLayer.prototype, 'updateToken', function (wrapped, ...args) {
			token = args[0];

			let result = wrapped.apply(this, args);

			token = null;

			return result;
		});

		new ResilientWrapper(SightLayer.prototype, 'updateFog', function (wrapped, ...args) {
			if(game.user.isGM && token != null && getSetting('fow-only-player-visible') && !token.playerObservable())
				return;

			return wrapped.apply(this, args);
		});
	})();


	new ResilientWrapper(SightLayer.prototype, 'update', (function() {
		const MAX_UVS_LENGTH = 1000;

		const hotfix_uv_lengths = function() {
			if(this.fog?.update && getSetting('fix-fow-memory-leak')) {
				if(this.fog.update.los.geometry.uvs.length > MAX_UVS_LENGTH || this.fog.update.fov.geometry.uvs.length > MAX_UVS_LENGTH)
					this._commitFogUpdate();
			}
		};

		const hotfix_update = function(wrapped, ...args) {
			let result =  wrapped.apply(this, args);

			hotfix_uv_lengths.apply(this);

			return result;
		};

		return async function(wrapped, ...args) {
			// We can skip the hook for non-GMs
			if(!game.user.isGM || !getSetting('fow-only-player-visible'))
				return hotfix_update.apply(this, arguments);


			// Swallow canvas.sight.fog.update.fov draws
			let fogUpdates = this._fogUpdates;
			this._fogUpdates = 0;

			// Call original method
			let result = wrapped.apply(this, args);

			// Restore fogUpdates
			this._fogUpdates = fogUpdates;

			// Manually do fog.update.fov / los draws
			// See original implementation in SightLayer.update
			if(!this.tokenVision || !this.sources.vision.size)
				return result;

			if(this._fogUpdates) {
				let fog = this.fog.update;
				let channels = this._channels;

				for ( let sources of Object.values(this.sources) ) {
					for ( let source of sources ) {
						if(source[0].startsWith('Token.')) {
							let token_id = source[0].slice(6);

							let token = canvas.tokens.get(token_id);

							if(!token || !token.playerObservable())
								continue;
						}

						let s = source[1];

						// Draw line of sight polygons
						if (s.los)
							fog.los.beginFill(0xFFFFFF, 1.0).drawPolygon(s.los).endFill();

						// Draw fog exploration polygons
						if((s.channels.dim + s.channels.bright) > 0)
							fog.fov.beginFill(channels.explored.hex, 1.0).drawPolygon(s.fov).endFill();
					}
				}

				hotfix_uv_lengths.apply(this);
			}

			// Done
			return result;
		};
	})());
});



token-vision-tweaks.js
// SPDX-License-Identifier: GPLv3-or-later
// Copyright © 2020 fvtt-token-vision-tweaks Rui Pinheiro

'use strict';

import {libWrapper} from './shim.js';

Hooks.once('ready', () => {
	const MODULE_NAME = "Token Vision Tweaks";
	const MODULE_ID = "token-vision-tweaks";

	if(isNewerVersion(game.data.version, "0.7.x")) {
		const MESSAGE = `
			<p>'${MODULE_NAME}' is no longer needed in Foundry 0.8 or newer, as the issues it attempts to resolve have been fixed.</p>
			<p><b>As a result, it will now disable itself.</b> If you do not plan on going back to Foundry 0.7, you should consider uninstalling it.</p>
		`;
		new Dialog({
			title: MODULE_NAME,
			content: MESSAGE, buttons: {
				ok: { icon: '<i class="fas fa-check"></i>', label: 'Understood' }
			}
		}).render(true);
		return;
	}


	const ORIGINAL_RAY_DENSITY = 6; // from SightLayer.computeSight
	const ORIGINAL_RAY_DEDUP_TOLERANCE = 50; // from SightLayer._castRays
	const ORIGINAL_EXACT_VISION_THRESHOLD = 500; // From SightLayer


	console.log(`Loading ${MODULE_NAME} module...`);


	//---------------------------
	// Settings
	const resetSight = function() {
		// Re-initialize vision sources in the scene
		canvas.initializeSources();
	};

	game.settings.register(MODULE_ID, 'ray-dedup-tolerance', {
		name: 'Ray De-duplication Tolerance',
		default: 0,
		type: Number,
		scope: 'world',
		scene: true,
		config: true,
		hint: `The tolerance used by Foundry when performing ray de-duplication. This option can be used to massively improve vision on large open maps at a cost of performance. The Foundry default is '${ORIGINAL_RAY_DEDUP_TOLERANCE}'.  Higher is better, but also slower. Disable this override with '0'.`,
		onChange: value => resetSight()
	});

	game.settings.register(MODULE_ID, 'ray-density', {
		name: 'Ray Density',
		default: 0,
		type: Number,
		scope: 'world',
		scene: true,
		config: true,
		hint: `The density in degrees to which to guarantee that rays are broadcast. This option can be used to slightly improve vision on large open maps at a cost of performance, although in most cases increasing the 'Ray De-duplication Tolerance' is much more effective. The Foundry default is '${ORIGINAL_RAY_DENSITY}'. Lower is better, but also slower. Disable this override with '0'.`,
		onChange: value => resetSight()
	});

	game.settings.register(MODULE_ID, 'max-radius', {
		name: 'Maximum Vision Distance (Grid Units)',
		default: 0,
		type: Number,
		scope: 'world',
		scene: true,
		config: true,
		hint: "Maximum token vision distance in grid units, regardless of illumination settings. This can be used to e.g. limit view distance when Global Illumination is turned on, or improve performance on very large open maps by limiting how far away tokens need to render sight. Disable this override with '0'.",
		onChange: value => resetSight()
	});

	game.settings.register(MODULE_ID, 'exact-vision-threshold', {
		name: 'Exact Vision Threshold',
		default: 0,
		type: Number,
		scope: 'world',
		scene: true,
		config: true,
		hint: `The number of walls starting at which vision becomes approximate. The Foundry default is '${ORIGINAL_EXACT_VISION_THRESHOLD}'.  Higher is better, but also slower. Disable this override with '0'.`,
		onChange: value => resetSight()
	});

	const getSetting = function(key) {
		let result = game.scenes.active.getFlag(MODULE_ID, key);

		if(result === null || result === undefined)
			result = game.settings.get(MODULE_ID, key);

		return result;
	};


	//---------------------------
	// Add settings to the SceneConfig dialog
	Hooks.on('renderSceneConfig', (app, html, data) => {
		const selector = html.find('header:contains(Vision)').nextAll('header')[0];
		console.log(selector);

		game.settings.settings.forEach(setting => {
			if(setting.module != MODULE_ID)
				return;

			if(!setting.scene)
				return;

			let value = app.object.getFlag(MODULE_ID, setting.key);
			if(value === undefined || value === null)
				value = "";

			$(`
				<div class="form-group">
					<label>${setting.name}</label>
					<input type="text" name="flags.${MODULE_ID}.${setting.key}" placeholder="${setting.type.name}" value="${value}" data-dtype="${setting.type.name}"/>
					<p class="notes">${setting.hint} Leave empty to use value set in the ${MODULE_NAME} module settings.</p>
				</div>
			`).insertBefore(selector);
		});
	});


	Hooks.on('updateScene', (entity, diff, options, user_id) => {
		if(!diff.flags)
			return;

		if(!diff.flags[MODULE_ID])
			return;

		resetSight();
	});




	//---------------------------
	// Hook the SightLayer computeSight method and implement the ray density / max radius module functionality
	libWrapper.register(MODULE_ID, 'SightLayer.computeSight', function(computeSight, ...args) {
		let options = args[2];

		// Modify exact vision threshold if necessary
		let exactVisionThreshold = getSetting('exact-vision-threshold');
		let previousExactVisionThreshold = SightLayer.EXACT_VISION_THRESHOLD;
		if(exactVisionThreshold != 0)
			SightLayer.EXACT_VISION_THRESHOLD = exactVisionThreshold;

		// Ray Density
		let rayDensity = getSetting('ray-density');

		if(rayDensity && rayDensity > 0)
			options.density = rayDensity;


		// Sight Radius
		let maxRadius = (getSetting('max-radius') * canvas.dimensions.size) / canvas.dimensions.distance;

		if(maxRadius && maxRadius > 0)
			args[1] = Math.min(args[1], maxRadius);

		// Call wrapped method
		const result = computeSight.apply(this, args);

		// Restore old exact vision threshold
		if(exactVisionThreshold != 0)
			SightLayer.EXACT_VISION_THRESHOLD = previousExactVisionThreshold;

		// Done
		return result;
	}, 'WRAPPER');


	//---------------------------
	// Hook the SightLayer _castRays method and implement the ray dedup tolerance functionality
	libWrapper.register(MODULE_ID, 'SightLayer._castRays', function(originalCastRays, ...args) {
		// We transparently wrap if the ray dedup tolerance patch is disabled
		const tol = getSetting('ray-dedup-tolerance');
		if(tol == 0)
			return originalCastRays(...args);

		// Grab the parameters
		const x = args[0];
		const y = args[1];
		const distance = args[2];
		const options  = args[3];
		const density = options.density ?? 4;
		const endpoints = options.endpoints;
		const limitAngle = options.limitAngle ?? false;
		const aMin = options.aMin;
		const aMax = options.aMax;


		// ---- Implementation copied from SightLayer._castRays of Foundry 0.7.5
		const rOffset = 0.02;

		// Enforce that all rays increase in angle from minimum towards maximum
		const rMin = limitAngle ? Ray.fromAngle(x, y, aMin, distance) : null;
		const rMax = limitAngle ? Ray.fromAngle(x, y, aMax, distance) : null;

		// Define de-duping cast function
		const cast = (ray) => { /* <= MODULE CUSTOMIZATION */
			let a = Math.round(ray.angle * tol) / tol;
			if ( angles.has(a) ) return;
			rays.push(ray);
			angles.add(a);
		};

		// Track rays and unique emission angles
		const angles = new Set();
		const rays = [];

		// First prioritize rays which are cast directly at wall endpoints
		for ( let e of endpoints ) {
			const ray = Ray.fromAngle(x, y, Math.atan2(e[1]-y, e[0]-x), distance);
			if ( limitAngle ) {
				ray.angle = this._adjustRayAngle(aMin, ray.angle);  // Standardize the angle
				if (!Number.between(ray.angle, aMin, aMax)) continue;
			}
			cast(ray);
		}

		// Next cast rays at any non-duplicate offset angles
		const nr = rays.length;
		for ( let i=0; i<nr; i++ ) {
			const r = rays[i];
			cast(r.shiftAngle(rOffset));
			cast(r.shiftAngle(-rOffset));
		}

		// Add additional limiting and central rays
		if ( limitAngle ) {
			const aCenter = aMin + ((aMax - aMin) / 2) + Math.PI;
			const rCenter = Ray.fromAngle(x, y, aCenter, 0);
			rCenter._isCenter = true;
			cast(rMin);
			cast(rCenter);
			cast(rMax);
		}

		// Add additional approximate rays to reach a desired radial density
		if ( !!density ) {
			const rDensity = toRadians(density);
			const nFill = Math.ceil((aMax - aMin) / rDensity);
			for ( let a of Array.fromRange(nFill) ) {
				cast(Ray.fromAngle(x, y, aMin + (a * rDensity), distance), 10);
			}
		}

		// Sort rays counter-clockwise (increasing radians)
		rays.sort((r1, r2) => r1.angle - r2.angle);
		return rays;
	});
});

